Fix symlink reading

RFC1920 opened the pull request:
StringPool.Take was always returning a size/count of 16.  So, any symlink target longer than 16 characters was being truncated.  Hard-coded to 256 as a max path, allowing the plugin watcher to work for symlinks.

This 256 character limit might be a problem for some, but 16 is generally too small.  In other words, there is probably a better way to handle this.

Also added a catch to the try when reading symlinks.  Without this, it was blocking server load at the first error encountered randomly with different linked plugins.

Together, these allow my server to load, with all linked plugins, and changes made to the plugins cause immediate reload of the plugin as with non-linked plugins.

groupwhere committed:
Add catch to prevent block to server startup upon failed reading of linked plugin.

groupwhere committed:
Restore

RFC1920 committed:
Update FSWatcher.cs to read full symlink path

StringPool.Take was always returning a size/count of 16.  So, any symlink target longer than 16 characters was being truncated.  Hard-coded to 256 as a max path, allowing the plugin watcher to work for symlinks.

austinv900 commented on src/Plugins/Watchers/FSWatcher.cs:
```diff
@@ -85,7 +85,7 @@ private bool IsFileSymlink(string path)
         private void LoadWatcherSymlink(string path)
         {
-            StringBuilder str = StringPool.Take();
+            StringBuilder str = new StringBuilder() { Capacity = 256 };
```
The only issue here is this creates garbage which is why we pool objects.

You'll want to pull from the pool and manually set the capacity if it ain't already the desired length.

RFC1920 replied:
I will look into how to do that.  Basically, without it it does not auto resize to the path length of the target and simply truncates it.  So, the code completes but no usable watch is set since the path we are watching doesn't exist.

austinv900 commented:
I think I had it like this at one point but continued to have issues. This is the proper way of handling things.

Windows and Linux provide two different path lengths and windows also offers long paths which is also a different size. We'd probably have to introduce a config variable since .NET doesn't make it easy to get the actual value. 

I'm curious to know how often a exception is being thrown.

RFC1920 commented:
Without the catch, on my system (AlmaLinux 8), the server would crash every single time during boot.  However, the catch may be unnecessary given that the real issue appears to have been the buffer size getting capped at 16.  I can't explain why even with the catch throwing errors that all plugins appeared to be loaded after server startup was completed.

RFC1920 commented:
Also, note that afaict, this only gets called for Unix and not Windows.  So, a PATH_MAX of 4096 is in play.  Unlikely, but that is a typical max.

RFC1920 commented:
Switching back to this:
```cs
StringBuilder str = StringPool.Take();
str.Capacity = 4096;
```
with the finally re-added appears to work, and no exception was logged.  So, it may be that we do not need the catch but rather enough buffer to handle the target path.

RFC1920 committed:
Roll back to simpler fix for stringpool issue

Setting capacity on the string pool variable appears to be enough.  Removed the catch and reverted back to stringpool.take with a capacity expansion to typical PATH_MAX of 4096.  Using count for length again as discovered.
